lib/gosu/color.rb

```diff
@@ -0,0 +1,175 @@
+      Gosu::Color.new(255, r, g, b)
+    end
+
+    def self.from_ahsv(alpha, h, s, v)
+      __color = Gosu_Color_create_from_ahsv(alpha.clamp(0.0, 255.0), h % 360.0, s.clamp(0.0, 1.0), v.clamp(0.0, 1.0))
+      Gosu.check_last_error
+      Gosu::Color.new(__color)
+    end
+
+    def self.from_hsv(h, s, v)
+      __color = Gosu_Color_create_from_hsv(h % 360.0, s.clamp(0.0, 1.0), v.clamp(0.0, 1.0))
+      Gosu.check_last_error
+      Gosu::Color.new(__color)
+    end
+
+    def initialize(*args)
+      case args.size
+      when 1
+        @__color = Gosu_Color_create(args.first)
+        Gosu.check_last_error
+      when 4
+        @__color = Gosu_Color_create_argb(args[0].clamp(0, 255), args[1].clamp(0, 255), args[2].clamp(0, 255), args[3].clamp(0, 255))
+        Gosu.check_last_error
+      else
+        raise ArgumentError
+      end
+    end
+
+    def alpha
+      Gosu_Color_alpha(@__color).tap { Gosu.check_last_error }
+    end
+
+    def alpha=(value)
+      @__color = Gosu_Color_set_alpha(@__color, value.clamp(0, 255)).tap { Gosu.check_last_error }
+    end
+
+    def red
+      Gosu_Color_red(@__color).tap { Gosu.check_last_error }
+    end
+
+    def red=(value)
+      @__color = Gosu_Color_set_red(@__color, value.clamp(0, 255)).tap { Gosu.check_last_error }
+    end
+
+    def green
+      Gosu_Color_green(@__color).tap { Gosu.check_last_error }
+    end
+
+    def green=(value)
+      @__color = Gosu_Color_set_green(@__color, value.clamp(0, 255)).tap { Gosu.check_last_error }
+    end
+
+    def blue
+      Gosu_Color_blue(@__color).tap { Gosu.check_last_error }
+    end
+
+    def blue=(value)
+      @__color = Gosu_Color_set_blue(@__color, value.clamp(0, 255)).tap { Gosu.check_last_error }
+    end
+
+    def value
+      Gosu_Color_value(@__color).tap { Gosu.check_last_error }
+    end
+
+    def value=(value)
+      @__color = Gosu_Color_set_value(@__color, value.clamp(0.0, 1.0)).tap { Gosu.check_last_error }
+    end
+
+    def saturation
+      Gosu_Color_saturation(@__color).tap { Gosu.check_last_error }
+    end
+
+    def saturation=(value)
+      @__color = Gosu_Color_set_saturation(@__color, value.clamp(0.0, 1.0)).tap { Gosu.check_last_error }
+    end
+
+    def hue
+      Gosu_Color_hue(@__color).tap { Gosu.check_last_error }
+    end
+
+    def hue=(value)
+      @__color = Gosu_Color_set_hue(@__color, value % 360.0).tap { Gosu.check_last_error }
+    end
+
+    def bgr
+      Gosu_Color_bgr(@__color).tap { Gosu.check_last_error }
+    end
+
+    def abgr
+      Gosu_Color_abgr(@__color).tap { Gosu.check_last_error }
+    end
+
+    def argb
+      Gosu_Color_argb(@__color).tap { Gosu.check_last_error }
+    end
+
+    def gl
+      @__color
+    end
+
+    def ==(other)
+      gl == (other.is_a?(Numeric) ? other : other&.gl)
+    end
+
+    def <(other)
+      gl < (other.is_a?(Numeric) ? other : other&.gl)
+    end
+
+    def >(other)
+      gl > (other.is_a?(Numeric) ? other : other&.gl)
+    end
```

Comment:
Can these be delegated to <=> as in C++20?

Reply:
The Ruby module Comparable does the trick!

Reply:
Thanks, done.